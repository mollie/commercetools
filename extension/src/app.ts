import express, { Application } from 'express';
import handler from '../expressHandler';
import morganMiddleware from './logger/morganMiddleware';

const app: Application = express();

app.use(morganMiddleware); // Logs from http level

app.get('/health', (req, res) => {
  res.send('Ok');
});

app.use(express.json());

app.use('/', handler);

export default app;
