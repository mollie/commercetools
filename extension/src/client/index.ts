import createMollieClient, { MollieClient } from '@mollie/api-client';
import { ClientBuilder, Client, AuthMiddlewareOptions, HttpMiddlewareOptions } from '@commercetools/sdk-client-v2';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}

export function initialiseCommercetoolsClient(): Client {
  const {
    commercetools: { projectKey, clientId, clientSecret, host, authUrl, scopes },
  } = config;

  const authMiddlewareOptions: AuthMiddlewareOptions = {
    host: authUrl,
    projectKey,
    credentials: {
      clientId,
      clientSecret,
    },
    scopes,
    fetch,
  };

  const httpMiddlewareOptions: HttpMiddlewareOptions = {
    host,
    fetch,
  };

  const client: Client = new ClientBuilder()
    .withClientCredentialsFlow(authMiddlewareOptions)
    .withHttpMiddleware(httpMiddlewareOptions)
    .withUserAgentMiddleware()
    .build();

  return client;
}
