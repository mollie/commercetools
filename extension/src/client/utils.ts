import createMollieClient, { MollieClient } from '@mollie/api-client';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}
