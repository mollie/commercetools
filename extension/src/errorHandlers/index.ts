import { formatMollieErrorResponse } from './formatMollieErrorResponse';
import { formatExtensionErrorResponse } from './formatExtensionErrorResponse';

const errorHandler = {
  mollie: formatMollieErrorResponse,
  extension: formatExtensionErrorResponse,
};

export default errorHandler;
