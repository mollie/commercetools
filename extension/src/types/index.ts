// This is not exhaustive
// If you use another commercetools error response code, add it to this enum
export enum CTEnumErrors {
  General = 'General',
  InvalidInput = 'InvalidInput',
  InvalidOperation = 'InvalidOperation',
  Unauthorized = 'Unauthorized',
  SyntaxError = 'SyntaxError',
  SemanticError = 'SemanticError',
  ObjectNotFound = 'ObjectNotFound',
}

export type CTUpdatesRequestedResponse = {
  status: number;
  actions?: Action[];
  errors?: CTError[];
};

export type Action = {
  action: string;
  type?: {
    key: string;
  };
  fields?: {
    actionType: string;
    request?: string;
    response?: string;
    createdAt?: string;
  };
  name?: string;
  value?: string;
  key?: string;
  transaction?: CTTransaction;
  transactionId?: string;
  interactionId?: string;
  state?: CTTransactionState;
};

export type CTError = {
  code: CTEnumErrors;
  message: string;
  extensionExtraInfo?: Object;
};

export type CTMoney = {
  type?: 'centPrecision';
  currencyCode: string;
  centAmount: number;
  fractionDigits?: number;
};

export type CTPayment = {
  id: string;
  amountPlanned: CTMoney;
  transactions?: CTTransaction[];
  key?: string;
  custom?: {
    fields?: {
      paymentMethodsRequest?: string;
      paymentMethodsResponse?: string;
      createPayment?: string;
      createCapture?: string;
      cancelPayment?: string;
      createRefund?: string;
    };
  };
  interfaceInteractions?: CTInterfaceInteraction[];
};

export type CTInterfaceInteraction = {
  id: string;
  actionType: string;
  createdAt?: Date;
  request?: string;
  response?: string;
};

export type CTTransaction = {
  id?: string;
  timestamp?: string;
  type: CTTransactionType;
  amount: CTMoney;
  interactionId?: string;
  state?: 'Initial' | 'Pending' | 'Success' | 'Failure';
};

export enum ControllerAction {
  GetPaymentMethods = 'getPaymentMethods',
  CreateOrder = 'createOrder',
  CreateOrderPayment = 'createOrderPayment',
  CreateOrderRefund = 'createOrderRefund',
  CreateCustomRefund = 'createCustomRefund',
  CreateShipment = 'createShipment',
  UpdateShipment = 'updateShipment',
  CancelOrder = 'cancelOrder',
  NoAction = 'noAction',
  Error = 'error',
}

export enum CTTransactionType {
  Authorization = 'Authorization',
  CancelAuthorization = 'CancelAuthorization',
  Charge = 'Charge',
  Refund = 'Refund',
  Chargeback = 'Chargeback',
}

export enum CTTransactionState {
  Initial = 'Initial',
  Pending = 'Pending',
  Success = 'Success',
  Failure = 'Failure',
}
