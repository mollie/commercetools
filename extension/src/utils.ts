import { MethodsListParams } from '@mollie/api-client';
/**
 * Generates an ISO string date
 * @returns {String} Returns the current date converted to ISO.
 */
export function createDateNowString() {
  return new Date().toISOString();
}

export function amountMapper(amountPlanned: any): string {
  const { centAmount, fractionDigits } = amountPlanned;
  const divider = Math.pow(10, fractionDigits || 2);
  const mollieAmount = (centAmount / divider).toFixed(2);
  return mollieAmount
}

export function methodListMapper(ctObj: any): MethodsListParams {
  // Generally this shouldn't be needed, but a safety anyway.. eventually could return error here
  if (!ctObj.amountPlanned) {
    return {};
  }
  const mollieAmount = amountMapper(ctObj.amountPlanned);

  const mObject: MethodsListParams = {
    amount: {
      value: mollieAmount,
      currency: ctObj.amountPlanned.currencyCode,
    },
    // Resource is hardcoded, for the time being we only support Orders API
    resource: 'orders',
  };

  if (ctObj.custom?.fields?.paymentMethodsRequest) {
    const parsedMethodsRequest = JSON.parse(ctObj.custom?.fields?.paymentMethodsRequest)
    const { locale, billingCountry, includeWallets, orderLineCategories, issuers, pricing, sequenceType } = parsedMethodsRequest;
    const include = issuers || pricing ? `${issuers ? 'issuers,' : ''}${pricing ? 'pricing' : ''}` : undefined;

    Object.assign(
      mObject,
      locale && { locale: locale },
      include && { include: include },
      includeWallets && { includeWallets: includeWallets },
      billingCountry && { billingCountry: billingCountry },
      sequenceType && { sequenceType: sequenceType },
      orderLineCategories && { orderLineCategories: orderLineCategories },
    );
  }

  return mObject;
}
