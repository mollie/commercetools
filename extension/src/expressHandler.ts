import {Request, Response} from 'express';
import handleRequest from './requestHandlers/handleRequest';
import {HandleRequestInput, HandleRequestSuccess} from './types';
import actions from "./requestHandlers/actions";


export default async function handler(req: Request, res: Response) {
    const requestInput = new HandleRequestInput(req.path, req.method, req.body);
    const result = await handleRequest(requestInput);
    if (result instanceof HandleRequestSuccess) {
        return res.status(200).send({actions: actions});

    } else {
        return res.status(result.status).send({errors: result.errors});
    }
};
