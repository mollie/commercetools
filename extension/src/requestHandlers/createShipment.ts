import { isEmpty, trim } from 'lodash';
import { MollieClient, ShipmentCreateParams, Shipment, Order, OrderLine, OrderLineType } from '@mollie/api-client';
import { v4 as uuid } from 'uuid';
import formatErrorResponse from '../errorHandlers';
import { Action, ControllerAction, CTPayment, CTTransaction, CTTransactionState, CTTransactionType, CTUpdatesRequestedResponse } from '../types';
import Logger from '../../src/logger/logger';
import { makeActions } from '../makeActions';
import { makeMollieAmount } from '../utils';

export function getShipmentParams(ctPayment: Required<CTPayment>, mollieOrder: Order | undefined): Promise<ShipmentCreateParams> {
  try {
    const shipmentParams: ShipmentCreateParams = {
      orderId: ctPayment.key,
    };
    if (isPartialCapture(ctPayment.transactions) && mollieOrder) {
      const initialCharge = findInitialCharge(ctPayment.transactions);
      const mollieLines = ctToMollieLines(initialCharge!, mollieOrder.lines);
      Object.assign(shipmentParams, { lines: mollieLines });
    }
    Logger.debug('shipmentParams: %o', shipmentParams);
    return Promise.resolve(shipmentParams);
  } catch (error) {
    Logger.error({ error });
    return Promise.reject({ status: 400, title: 'Could not make parameters needed to create Mollie shipment.', field: 'createShipmentRequest' });
  }
}

function tryParseJSON(jsonString: string | undefined) {
  try {
    const parsed = JSON.parse(jsonString!);
    if (parsed && typeof parsed === 'object') return parsed;
  } catch (error) {
    return false;
  }
}

export function ctToMollieLines(ctTransaction: CTTransaction, mollieOrderLines: OrderLine[]): Object[] {
  const parsedOptions = tryParseJSON(ctTransaction.custom?.fields?.lineIds);
  const ctLinesArray = parsedOptions ? parsedOptions : ctTransaction.custom?.fields?.lineIds?.split(',').map(trim);

  const mollieLines = ctLinesArray.reduce((acc: Object[], ctLine: any) => {
    const ctLineId = typeof ctLine === 'string' ? ctLine : ctLine.id;
    const mollieLine = ctLineId && mollieOrderLines.find(mollieLine => mollieLine.metadata?.cartLineItemId === ctLineId || mollieLine.metadata?.cartCustomLineItemId === ctLineId);
    if (mollieLine) {
      const transformedLine = { id: mollieLine.id };
      ctLine.quantity && Object.assign(transformedLine, { quantity: ctLine.quantity });
      ctLine.totalPrice && Object.assign(transformedLine, { amount: makeMollieAmount(ctLine.totalPrice) });
      acc.push(transformedLine);
    }
    return acc;
  }, []);

  if (ctTransaction.custom?.fields?.includeShipping) {
    const shippingLine = mollieOrderLines.find(mollieLine => mollieLine.type === OrderLineType.shipping_fee);
    shippingLine && mollieLines.push({ id: shippingLine.id });
  }

  return mollieLines;
}

export function mollieToCtLines(mollieOrderLines: OrderLine[]): string {
  const ctLinesString = mollieOrderLines.reduce((acc: string, orderLine: OrderLine) => {
    const ctLineId = orderLine.metadata?.cartLineItemId || orderLine.metadata?.cartCustomLineItemId;
    if (ctLineId) acc += `${ctLineId},`;
    if (orderLine.type === OrderLineType.shipping_fee) acc += `${orderLine.name},`;
    return acc;
  }, '');

  return ctLinesString;
}

export function findInitialCharge(transactions: CTTransaction[]): CTTransaction | undefined {
  // Assumes one initial transaction, i.e. one capture being made at a time
  return transactions.find(tr => tr.type === CTTransactionType.Charge && tr.state === CTTransactionState.Initial);
}

export function isPartialCapture(transactions: CTTransaction[]): boolean {
  if (!transactions) return false;
  const initialCharge = findInitialCharge(transactions);
  return !isEmpty(initialCharge?.custom?.fields?.lineIds) || initialCharge?.custom?.fields?.includeShipping!;
}

export function createCtActions(mollieShipmentRes: Shipment, ctPayment: CTPayment): Action[] {
  const initialChargeTransaction = findInitialCharge(ctPayment.transactions!);
  const inTransactionId = initialChargeTransaction!.id || '';
  const mollieCreatedAt = mollieShipmentRes.createdAt;
  const interfaceInteractionId = uuid();
  const interfaceInteractionRequest = {
    transactionId: inTransactionId,
    createShipment: initialChargeTransaction?.custom?.fields,
  };
  const interfaceInteractionResponse = {
    mollieShipmentId: mollieShipmentRes.id,
    lineIds: mollieToCtLines(mollieShipmentRes.lines),
  };
  const interfaceInteractionParams = {
    id: interfaceInteractionId,
    actionType: ControllerAction.CreateShipment,
    requestValue: JSON.stringify(interfaceInteractionRequest),
    responseValue: JSON.stringify(interfaceInteractionResponse),
    timestamp: mollieCreatedAt,
  };

  const result: Action[] = [
    makeActions.changeTransactionState(inTransactionId, CTTransactionState.Success),
    makeActions.changeTransactionTimestamp(inTransactionId, mollieCreatedAt),
    makeActions.changeTransactionInteractionId(inTransactionId, interfaceInteractionId),
    makeActions.setStatusInterfaceText('Shipping'),
    makeActions.addInterfaceInteraction(interfaceInteractionParams),
  ];

  return result;
}

export default async function createShipment(ctPayment: Required<CTPayment>, mollieClient: MollieClient): Promise<CTUpdatesRequestedResponse> {
  try {
    Logger.debug('ctPayment: %o', ctPayment);
    const mollieOrderRes = isPartialCapture(ctPayment.transactions ?? []) ? await mollieClient.orders.get(ctPayment.key) : undefined;
    const shipmentParams = await getShipmentParams(ctPayment, mollieOrderRes);
    const mollieShipmentRes = await mollieClient.orders_shipments.create(shipmentParams);
    Logger.debug('mollieShipmentRes: %o', mollieShipmentRes);
    const ctActions = createCtActions(mollieShipmentRes, ctPayment);
    return {
      actions: ctActions,
      status: 201,
    };
  } catch (error: any) {
    Logger.error({ error });
    const errorResponse = formatErrorResponse(error);
    return errorResponse;
  }
}
