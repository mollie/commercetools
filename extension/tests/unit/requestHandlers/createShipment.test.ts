import { mocked } from 'ts-jest/utils';
import { Action, CTPayment, CTShipment } from '../../../src/types';
import createShipment, { getShipmentParams, createCtActions } from '../../../src/requestHandlers/createShipment';
import { createDateNowString } from '../../../src/utils';
import Logger from '../../../src/logger/logger';

jest.mock('../../../src/utils');

describe('getShipmentParams', () => {
  const mockLogError = jest.fn();
  beforeEach(() => {
    Logger.error = mockLogError;
    mocked(createDateNowString).mockReturnValue('2021-10-08T12:12:02.625Z');
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('Should create required params for mollie createShipment call', async () => {
    const mockedCtObj = {
      key: 'ord_3uwvfd',
      custom: {
        fields: {
          createCapture: '{}',
        },
      },
    };
    const expectedCreateShipmentParams = {
      orderId: 'ord_3uwvfd',
    };
    await expect(getShipmentParams(mockedCtObj)).resolves.toEqual(expectedCreateShipmentParams);
  });
  it('Should create all optional params for mollie createShipment call', async () => {
    const mockedCtObj = {
      key: 'ord_3uwvfd',
      custom: {
        fields: {
          createCapture: '{"lines":[{"id":"testCartLineItemId","quantity":3}]}',
        },
      },
    };
    const expectedCreateShipmentParams = {
      orderId: 'ord_3uwvfd',
      lines: [{ id: 'testCartLineItemId', quantity: 3 }],
    };
    await expect(getShipmentParams(mockedCtObj)).resolves.toEqual(expectedCreateShipmentParams);
  });
  it('Should return 400 and error message if creating the parameters throws an error', async () => {
    const mockedCtObj = {
      key: 'ord_3uwvfd',
      custom: {
        fields: {
          createShipmentRequest: '',
        },
      },
    };
    const expectedRejectedValue = {
      status: 400,
      title: 'Could not make parameters needed to create Mollie shipment.',
      field: 'createShipmentRequest',
    };
    await expect(getShipmentParams(mockedCtObj)).rejects.toEqual(expectedRejectedValue);
    expect(mockLogError).toHaveBeenCalledTimes(1);
  });
});
describe('createCtActions', () => {
  beforeEach(() => {
    mocked(createDateNowString).mockReturnValue('2021-10-08T12:12:02.625Z');
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('Should create correct ct actions from request and mollies payment', () => {
    const mockedCtObject = {
      custom: {
        fields: {
          createCapture: '{"lines":[{"id":"odl_1.d8ck99","quantity":1}]}',
        },
      },
    };
    const mockedShipmentResponse: any = {
      resource: 'shipment',
      id: 'shp_t72vlb',
      orderId: 'ord_qzwg9x',
      createdAt: '2021-10-27T10:25:24+00:00',
      lines: [
        {
          resource: 'orderline',
          id: 'odl_1.d8ck99',
          orderId: 'ord_qzwg9x',
          name: 'orange',
          sku: null,
          type: 'physical',
          status: 'completed',
          isCancelable: false,
          quantity: 1,
          vatRate: '20.00',
          createdAt: '2021-10-27T10:02:36+00:00',
        },
      ],
    };
    const ctActions = createCtActions(mockedShipmentResponse, mockedCtObject);
    ctActions.forEach(action => {
      expect(action).toMatchSnapshot();
    });
  });
});

describe('createShipment', () => {
  const mockLogError = jest.fn();
  beforeEach(() => {
    Logger.error = mockLogError;
    mocked(createDateNowString).mockReturnValue('2021-10-08T12:12:02.625Z');
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('Should prepare params, call mollie, handle response and return actions', async () => {
    const mockedShipmentParams = { orderId: 'ord_qzwg9x', lines: [{ id: 'odl_1.d8ck99', quantity: 1 }] };
    const mockedCtObject: CTShipment = {
      key: 'ord_qzwg9x',
      custom: {
        fields: {
          createCapture: '{"lines":[{"id":"odl_1.d8ck99","quantity":1}]}',
        },
      },
    };
    const mockedShipmentResponse: any = {
      resource: 'shipment',
      id: 'shp_t72vlb',
      orderId: 'ord_qzwg9x',
      createdAt: '2021-10-27T10:25:24+00:00',
      lines: [
        {
          resource: 'orderline',
          id: 'odl_1.d8ck99',
          orderId: 'ord_qzwg9x',
          name: 'orange',
          sku: null,
          type: 'physical',
          status: 'completed',
          isCancelable: false,
          quantity: 1,
          vatRate: '20.00',
          createdAt: '2021-10-27T10:02:36+00:00',
        },
      ],
    };
    const mollieClient = { orders_shipments: { create: jest.fn().mockResolvedValueOnce(mockedShipmentResponse) } } as any;

    const createShipmentRes = await createShipment(mockedCtObject, mollieClient);
    expect(mollieClient.orders_shipments.create).toHaveBeenCalledWith(mockedShipmentParams);
    expect(createShipmentRes.actions).toHaveLength(2);
    expect(createShipmentRes.status).toBe(201);
  });
  it('Should return commercetools formated error if one of the functions fails', async () => {
    const mockedError = { status: 400, title: 'Could not make parameters needed to create Mollie shipment.', field: 'createShipmentRequest' };
    // const getShipmentParams = jest.fn().mockRejectedValueOnce(mockedError);
    // const createCtActions = jest.fn();
    const mollieClient = { orders_shipments: { create: jest.fn() } } as any;

    const createShipmentRes = await createShipment({} as CTShipment, mollieClient);
    expect(createShipmentRes.status).toBe(400);
    expect(createShipmentRes.errors).toHaveLength(1);
  });
});
