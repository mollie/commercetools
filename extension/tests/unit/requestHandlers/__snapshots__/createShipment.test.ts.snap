// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCtActions Should create correct ct actions from request and mollies payment 1`] = `
Object {
  "action": "addInterfaceInteraction",
  "fields": Object {
    "actionType": "createShipment",
    "createdAt": "2021-10-08T12:12:02.625Z",
    "request": "{\\"lines\\":[{\\"id\\":\\"odl_1.d8ck99\\",\\"quantity\\":1}]}",
    "response": "{\\"resource\\":\\"shipment\\",\\"id\\":\\"shp_t72vlb\\",\\"orderId\\":\\"ord_qzwg9x\\",\\"createdAt\\":\\"2021-10-27T10:25:24+00:00\\",\\"lines\\":[{\\"resource\\":\\"orderline\\",\\"id\\":\\"odl_1.d8ck99\\",\\"orderId\\":\\"ord_qzwg9x\\",\\"name\\":\\"orange\\",\\"sku\\":null,\\"type\\":\\"physical\\",\\"status\\":\\"completed\\",\\"isCancelable\\":false,\\"quantity\\":1,\\"vatRate\\":\\"20.00\\",\\"createdAt\\":\\"2021-10-27T10:02:36+00:00\\"}]}",
  },
  "type": Object {
    "key": "ct-mollie-integration-interface-interaction-type",
  },
}
`;

exports[`createCtActions Should create correct ct actions from request and mollies payment 2`] = `
Object {
  "action": "setCustomField",
  "name": "createShipmentResponse",
  "value": "{\\"resource\\":\\"shipment\\",\\"id\\":\\"shp_t72vlb\\",\\"orderId\\":\\"ord_qzwg9x\\",\\"createdAt\\":\\"2021-10-27T10:25:24+00:00\\",\\"lines\\":[{\\"resource\\":\\"orderline\\",\\"id\\":\\"odl_1.d8ck99\\",\\"orderId\\":\\"ord_qzwg9x\\",\\"name\\":\\"orange\\",\\"sku\\":null,\\"type\\":\\"physical\\",\\"status\\":\\"completed\\",\\"isCancelable\\":false,\\"quantity\\":1,\\"vatRate\\":\\"20.00\\",\\"createdAt\\":\\"2021-10-27T10:02:36+00:00\\"}]}",
}
`;
